hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcOutputStream.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.container.replication;
+
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos.CopyContainerResponseProto;
+import org.apache.ratis.thirdparty.com.google.protobuf.ByteString;
+import org.apache.ratis.thirdparty.io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * Adapter from {@code OutputStream} to gRPC {@code StreamObserver}.
+ * Data is buffered in a limited buffer of the specified size.
+ */
+class GrpcOutputStream extends OutputStream {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(GrpcOutputStream.class);
+
+  private final StreamObserver<CopyContainerResponseProto> responseObserver;
+
+  private final ByteString.Output buffer;
+
+  private final long containerId;
+
+  private final int bufferSize;
+
+  private int writtenBytes;
+
+  GrpcOutputStream(
+      StreamObserver<CopyContainerResponseProto> responseObserver,
+      long containerId, int bufferSize) {
+    this.responseObserver = responseObserver;
+    this.containerId = containerId;
+    this.bufferSize = bufferSize;
+    buffer = ByteString.newOutput(bufferSize);
+  }
+
+  @Override
+  public void write(int b) {
+    try {
+      if (buffer.size() >= bufferSize) {
+        flushBuffer(false);
+      }
+      buffer.write(b);
```

Comment:
It should be write and then flush.  Otherwise, the data will be sitting in the buffer after write even the buffer is full.

Reply:
Fixed, and added test case.  Thanks.